### PR TITLE
Get built zip file name from project's metadata.

### PR DESCRIPTION
**Pull request by facundobatista**

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1,3 @@
+include README.md
```
Comment (chipaca) [inline MANIFEST.in:1]: README.md is included by default in new enough setuptools (the one in xenial is too old -- it doesn't support markdown at all afaict).

File: MANIFEST.in [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+include LICENSE
```
Comment (chipaca) [inline MANIFEST.in:2]: why is this needed? the metadata already says which license it is, we don't need to include the full text surely?

File: MANIFEST.in [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
+include requirements.txt
```
Comment (chipaca) [inline MANIFEST.in:3]: this will already be included in the egg-info

**Comment by facundobatista** [comment]
General comment about `MANIFEST.in`:

- `LICENSE`: it probably doesn't harm to distribute it
- `requirements.txt`: distributing it makes `setup.py` simpler because it needs to get the dependencies from one place only
- `README.md`: now that we have the MANIFEST.in file, it's better to be explicit

Also, we may have some other files in the future that would make sense to include here for distribution (e.g. docs).

File: MANIFEST.in (hunk 1 above)
Comment (jameinel) [reply MANIFEST.in:1]: Just to clarify, this is the MANIFEST of charmcraft itself, right? not of the charms that it builds?

File: MANIFEST.in (hunk 2 above)
Comment (jameinel) [reply MANIFEST.in:2]: https://www.gnu.org/licenses/gpl-howto.html
says
"You should also include a copy of the license itself somewhere in the distribution of your program. All programs, whether they are released under the GPL or LGPL, should include the text version of the GPL."

I do believe that the packaged version of charmcraft counts as the "distribution" of the program, and thus it *should* include the full text of the LICENSE that it is using. (GNU also happens to use COPYING rather than LICENSE, but I don't particularly care about the name as much as 'should you include a copy' seems to bias that it should be part of the program.)

File: charmcraft/commands/build.py [hunk 4]
```diff
@@ -175,8 +177,12 @@ def handle_dependencies(self):
 
     def handle_package(self):
         """Handle the final package creation."""
+        logger.debug("Parsing the project's metadata")
+        with (self.charmdir / CHARM_METADATA).open('rt', encoding='utf8') as fh:
+            metadata = yaml.safe_load(fh)
+
         logger.debug("Creating the package itself")
-        zipname = self.charmdir.name + '.charm'
+        zipname = metadata['name'] + '.charm'
```
Comment (jameinel) [inline charmcraft/commands/build.py:185]: I definitely agree that the charm name should come from the metadata name and not the directory. That is the name that Juju will deploy it as.

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+PyYAML == 5.3.1
```
Comment (jameinel) [inline requirements.txt:1]: I'm surprised that these are exact versions rather than at least some sort of range (>5.3).

File: tests/commands/test_build.py [hunk 6]
```diff
@@ -360,10 +361,13 @@ def test_build_basic_complete_structure(tmp_path):
...
 
-    # the metadata
-    metadata = tmp_path / 'metadata.yaml'
-    with metadata.open('wb') as fh:
-        fh.write(b'lot of yaml config')
+    # the metadata (save it and restore to later check)
+    metadata_data = {'name': 'name-from-metadata'}
+    metadata_file = tmp_path / 'metadata.yaml'
+    with metadata_file.open('wt', encoding='ascii') as fh:
+        yaml.dump(metadata_data, fh)
```
Comment (jameinel) [inline tests/commands/test_build.py:368]: It seems a little odd to write it out and then read it again. Why not:
```
with metadata_file.open(...) as fh:
  raw = yaml.dump(metadata_data)
  fh.write(raw)
```

File: MANIFEST.in (hunk 2 above)
Comment (chipaca) [reply MANIFEST.in:2]: we're apache-licensed, but indeed it also says

>      (a) You must give any other recipients of the Work or
>          Derivative Works a copy of this License; and

File: MANIFEST.in (hunk 1 above)
Comment (facundobatista) [reply MANIFEST.in:1]: Yeap.

File: MANIFEST.in (hunk 2 above)
Comment (facundobatista) [reply MANIFEST.in:2]: Perfect, so we need this.

File: requirements.txt (hunk 5 above)
Comment (facundobatista) [reply requirements.txt:1]: If for some reason we break with 5.4, we don't want anybody installing the tool in the future to use 5.4. Remember this is a tool that "it should just work", not a library you're incorporating to your project (which you may run tests on, adjust versions, etc.)

File: tests/commands/test_build.py (hunk 6 above)
Comment (facundobatista) [reply tests/commands/test_build.py:368]: Ah, didn't know about that. I always wondering why `yaml` didn't have a `dumps` function.

Thanks! Improved.
